File: sdk/langchain/src/tools.ts
/**
 * Tool wrapper for ProceedGate cost governance
 */

import type { StructuredToolInterface } from '@langchain/core/tools';
import type {
  ProceedGateToolWrapperOptions,
  CheckRequest,
  CheckResponse,
  FrictionInfo,
} from './types.js';

export interface ProceedGateToolWrapper {
  wrap<T extends StructuredToolInterface>(tool: T): T;
}

/**
 * Wrap a LangChain tool with ProceedGate cost governance
 */
export function wrapToolWithGate<T extends StructuredToolInterface>(
  tool: T,
  options: ProceedGateToolWrapperOptions
): T {
  const { apiKey, workspaceId, baseUrl = 'https://governor.proceedgate.dev', policyId = 'retry_friction_v1', onFriction = 'warn' } = options;

  let attemptCount = 0;

  // Store original invoke method
  const originalInvoke = tool.invoke.bind(tool);

  // Override the invoke method
  (tool as unknown as { invoke: typeof originalInvoke }).invoke = async (input: unknown, config?: unknown) => {
    attemptCount++;

    const request: CheckRequest = {
      policy_id: policyId,
      action: 'tool_call',
      actor: {
        id: `langchain-tool:${workspaceId}`,
        project: workspaceId,
      },
      context: {
        attempt_in_window: attemptCount,
        window_seconds: 30,
        tool: tool.name,
        confidence: 0.9,
      },
    };

    const response = await fetch(`${baseUrl}/v1/governor/check`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`,
      },
      body: JSON.stringify(request),
    });

    const data = (await response.json()) as CheckResponse;

    if (!data.allowed) {
      const info: FrictionInfo = {
        decisionId: data.decision_id,
        reasonCode: data.reason_code,
        price: data.policy.friction_price,
        action: 'tool_call',
        tool: tool.name,
      };

      if (onFriction === 'block') {
        throw new Error(`ProceedGate: Tool ${tool.name} blocked. Friction required: ${info.price}`);
      }

      if (onFriction === 'warn') {
        console.warn(`[ProceedGate] Tool ${tool.name} requires friction: ${info.price}`);
      } else {
        const shouldContinue = await onFriction(info);
        if (!shouldContinue) {
          throw new Error(`ProceedGate: Tool ${tool.name} blocked by custom handler`);
        }
      }
    }

    // Reset attempt count on success path
    if (data.allowed) {
      attemptCount = 0;
    }

    return originalInvoke(input, config as Parameters<typeof originalInvoke>[1]);
  };

  return tool;
}

/**
 * Create a reusable tool wrapper
 */
export function createToolWrapper(options: ProceedGateToolWrapperOptions): ProceedGateToolWrapper {
  return {
    wrap<T extends StructuredToolInterface>(tool: T): T {
      return wrapToolWithGate(tool, options);
    },
  };
}

File: sdk/crewai/src/proceedgate_crewai/tools.py
"""
ProceedGate Tool Decorator for CrewAI
"""

from typing import Callable, Optional, TypeVar, Any
from functools import wraps
from .client import ProceedGateClient
from .exceptions import FrictionRequiredError

F = TypeVar("F", bound=Callable[..., Any])


class GatedTool:
    """
    Wrapper class for tools with ProceedGate cost governance.
    """
    
    def __init__(
        self,
        func: Callable,
        api_key: str,
        workspace_id: str,
        base_url: str = "https://governor.proceedgate.dev",
        policy_id: str = "retry_friction_v1",
        on_friction: str = "warn",
    ):
        self.func = func
        self.client = ProceedGateClient(api_key, workspace_id, base_url)
        self.policy_id = policy_id
        self.on_friction = on_friction
        self.attempt_count = 0
        
        # Preserve function metadata
        self.__name__ = func.__name__
        self.__doc__ = func.__doc__
    
    def __call__(self, *args, **kwargs):
        self.attempt_count += 1
        
        result = self.client.check(
            action="tool_call",
            policy_id=self.policy_id,
            context={
                "attempt_in_window": self.attempt_count,
                "window_seconds": 30,
                "tool": self.__name__,
            },
        )
        
        if not result.allowed:
            price = result.policy.get("friction_price", "unknown")
            
            if self.on_friction == "block":
                raise FrictionRequiredError(
                    f"Tool '{self.__name__}' blocked. Friction required: {price}",
                    decision_id=result.decision_id,
                    reason_code=result.reason_code or "friction",
                    price=price,
                    action="tool_call",
                )
            
            print(f"[ProceedGate] Warning: Tool '{self.__name__}' requires friction: {price}")
        
        # Reset on success
        if result.allowed:
            self.attempt_count = 0
        
        return self.func(*args, **kwargs)


def gated_tool(
    api_key: str,
    workspace_id: str,
    base_url: str = "https://governor.proceedgate.dev",
    policy_id: str = "retry_friction_v1",
    on_friction: str = "warn",
) -> Callable[[F], F]:
    """
    Decorator to wrap a function with ProceedGate cost governance.
    
    Usage:
        @gated_tool(api_key="...", workspace_id="...")
        def my_expensive_tool(query: str) -> str:
            return external_api_call(query)
    """
    def decorator(func: F) -> F:
        wrapper = GatedTool(
            func=func,
            api_key=api_key,
            workspace_id=workspace_id,
            base_url=base_url,
            policy_id=policy_id,
            on_friction=on_friction,
        )
        return wrapper  # type: ignore
    
    return decorator
